fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/PreOrderPruningIterator.java

```diff
@@ -48,13 +46,11 @@
 public final class PreOrderPruningIterator<T extends TreeLike<T>> implements Iterator<T> {
 
     @Nonnull
-    private final Deque<Pair<Iterable<? extends T>, Integer>> stack;
+    private final Deque<Iterator<? extends  T>> stack;
```

Comment:
The changes here was to convert this from a `Deque` of `Pair<Iterable, Integer>`, where the integer represented the current position within the `Iterable`, to a `Deque` of `Iterators`. Previously, in each iteration, we'd advance the `Integer` in the top `Pair` until we'd exhausted the tree. Now, we advance the `Iterator`.

For array-based `Iterable`s, the `Iterator` is more-or-less a `Pair<Iterable, Integer>`, and advancing the iterator works by getting the value in the array by position and then incrementing the value, so the implementations are basically the same. For linked list-based `Iterable`s, the `Iterator` effectively maintains a pointer to the correct node within the list, so this implementation would be a performance improvement (from quadratic to linear performance per node child size).

Reply:
That's a good improvement. By the way, at least in Cascades, we almost always `ImmutableList`, which is backed by an array, so the cost of `Iterables.get` is constant. Nevertheless, I think using an `iterator()` is more elegant that manually keeping track of the pointer.